### PR TITLE
autotools template: several improvements

### PR DESCRIPTION
**Pull request by SpaceIm**
- handle windows as build machine
- handle `tools.gnu:pkg_config` config
- handle msvc
- fix `install_name` of shared libs on macOS (to set `@rpath` instead of lib dir absolute path)
- call `VirtualBuildEnv` & `VirtualRunEnv` first in generate to avoid side effects

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Review by uilianries — APPROVED** [review]
Nice improvement, Windows support is a great addition! Thanks!

File: docs/package_templates/autotools_package/all/conanfile.py [hunk 1]
```diff
@@ -75,64 +85,99 @@ def requirements(self):
...
-        tc.configure_args.append("--enable-tools=no")
-        tc.configure_args.append("--enable-manpages=no")
+        # autotools usually uses 'yes' and 'no' to enable/disable options
+        yes_no = lambda v: "yes" if v else "no"
+        tc.configure_args.extend([
+            f"--with-foobar={yes_no(self.options.with_foobar)}",
+            "--enable-tools=no",
+            "--enable-manpages=no",
+        ])
         tc.generate()
```
Comment (jwillikers) [inline docs/package_templates/autotools_package/all/conanfile.py:131]: In #13355, I've just figured out that to get the compiler wrapper script to work, I had to pass in the Environment as an argument to the AutotoolsToolchain `generate()` function. Perhaps generating the toolchain after the Environment for the compiler wrapper has been set destroys the compiler wrapper variables. I think that's worth mentioning.

File: docs/package_templates/autotools_package/all/conanfile.py [hunk 2]
```diff
@@ -75,64 +85,99 @@ def requirements(self):
...
-            env.generate(scope="build")
+
+        # If Visual Studio is supported
+        if is_msvc(self):
+            env = Environment()
+            # get compile & ar-lib from automake (or eventually lib source code if available)
+            # it's not always required to wrap CC, CXX & AR with these scripts, it depends on how much love was put in
+            # upstream build files
+            compile_wrapper = unix_path(self, self._user_info_build["automake"].compile)
+            ar_wrapper = unix_path(self, self._user_info_build["automake"].ar_lib)
```
Comment (jwillikers) [inline docs/package_templates/autotools_package/all/conanfile.py:146]: We'll need to remember to update these soon when `automake` is updated to use the new `conf` variables.

**Review by jwillikers — APPROVED** [review]

**Review by jwillikers — APPROVED** [review]

File: docs/package_templates/autotools_package/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply docs/package_templates/autotools_package/all/conanfile.py:131]: Yes, the order in which generators are called my be important. Basically you want to call `Virtual[Build/Run]Env` first, then other generators, then specific `Environment`.

I've proposed https://github.com/conan-io/conan/pull/12194 for robust handling of Visual Studio in AutotoolToolchain, and honor in a robust way injection of CC, CXX etc from profile, but it deserves additional testing in conan client, and maybe a slightly different implementation (maybe a separate generator). It also suggest to add specific `conf` parameters to provide a canonical conan way to inject compiler path for users.

File: docs/package_templates/autotools_package/all/conanfile.py [hunk 3]
```diff
@@ -75,64 +85,99 @@ def requirements(self):
...
             raise ConanInvalidConfiguration(f"{self.ref} is not supported on {self.info.settings.os}.")
 
     # if another tool than the compiler or autotools is required to build the project (pkgconf, bison, flex etc)
     def build_requirements(self):
+        # only if we have to call autoreconf
         self.tool_requires("libtool/x.y.z")
-        self.tool_requires("pkgconf/x.y.z")
+        # only if upstream configure.ac relies on PKG_CHECK_MODULES macro
+        if not self.conf.get("tools.gnu:pkg_config", default=False, check_type=str):
+            self.tool_requires("pkgconf/x.y.z")
```
Comment (SpaceIm) [inline docs/package_templates/autotools_package/all/conanfile.py:99]: I've taken a look at conan source code, and conan v2 build helpers don't seem to do any special management of `tools.gnu:pkg_config`, so they don't guarantee that this path is discoverable.

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**